packages/standard-components/src/Search.svelte

```diff
@@ -0,0 +1,158 @@
+<script>
+  import { getContext } from "svelte"
+  import { isEmpty } from "lodash/fp"
+  import {
+    Button,
+    DatePicker,
+    Label,
+    Select,
+    Toggle,
+    Input,
+  } from "@budibase/bbui"
+
+  const { API, styleable, DataProvider, builderStore } = getContext("sdk")
+  const component = getContext("component")
+
+  export let table = []
+  export let columns = []
+  export let pageSize = 50
+  export let noRowsMessage = "No Rows"
+
+  let rows = []
+  let loaded = false
+  let search = {}
+  let tableDefinition
+  let schema
+
+  // pagination
+  let page = 0
+
+  $: fetchData(table, page)
+  // omit empty strings
+  $: parsedSearch = Object.keys(search).reduce(
+    (acc, next) =>
+      search[next] === "" ? acc : { ...acc, [next]: search[next] },
+    {}
+  )
+
+  async function fetchData(table, page) {
+    if (!isEmpty(table)) {
+      const tableDef = await API.fetchTableDefinition(table)
+      schema = tableDef.schema
+      rows = await API.searchTable({
+        tableId: table,
+        search: parsedSearch,
+        pagination: {
+          pageSize,
+          page,
+        },
+      })
+    }
+    loaded = true
+  }
+
+  function nextPage() {
+    page += 1
+  }
+
+  function previousPage() {
+    page -= 1
+  }
+</script>
+
+<div use:styleable={$component.styles}>
+  <div class="query-builder">
+    {#if schema}
+      {#each columns as field}
+        <div class="form-field">
+          <Label extraSmall grey>{schema[field].name}</Label>
+          {#if schema[field].type === 'options'}
+            <Select secondary bind:value={search[field]}>
+              <option value="">Choose an option</option>
+              {#each schema[field].constraints.inclusion as opt}
+                <option>{opt}</option>
+              {/each}
+            </Select>
+          {:else if schema[field].type === 'datetime'}
+            <DatePicker bind:value={search[field]} />
+          {:else if schema[field].type === 'boolean'}
+            <Toggle text={schema[field].name} bind:checked={search[field]} />
+          {:else if schema[field].type === 'number'}
+            <Input type="number" bind:value={search[field]} />
+          {:else if schema[field].type === 'string'}
+            <Input bind:value={search[field]} />
+          {/if}
+        </div>
+      {/each}
+    {/if}
+    <div class="actions">
+      <Button
+        secondary
+        on:click={() => {
+          search = {}
+          page = 0
+        }}>
+        Reset
+      </Button>
+      <Button primary on:click={() => fetchData(table)}>Search</Button>
+    </div>
+  </div>
+  {#if rows.length > 0}
+    {#if $component.children === 0 && $builderStore.inBuilder}
+      <p>Add some components too</p>
+    {:else}
+      {#each rows as row}
+        <DataProvider {row}>
+          <slot />
+        </DataProvider>
+      {/each}
+    {/if}
+  {:else if loaded && $builderStore.inBuilder}
+    <p>Feed me some data</p>
+  {:else}
+    <p>{noRowsMessage}</p>
```

Comment:
We should probably still check if `loaded` here so that we don't flash the no rows message before the initial load completes.

We also don't ever show the `noRowsMessage` in a repeater unless we're inside the builder, but with the search component it'll be shown in real apps. Should we standardise that? I hid it outside the builder for the repeater because it was unstyleable by users. Thinking outside the box here, but we could probably actually add an `EmptyState` or `NoRows` component which will only render itself if it's parent data source has no data? That would mean users could actually style and add any components they want to an empty state. That's outside the scope of this PR - I'm just thinking out loud about a future feature.

Reply:
I think that's a good idea - a lot of people like to show big banners with images and the like with "Not found" messages. It would be a nice UX to standardise that for Search and for the Repeater.

I felt it made more sense to show the `noRowsMessage` in real apps when using Search, because you are much more likely to hit that scenario, when your search doesn't return anything